Rewrite 'Regenerating configure' section of building guide (#1109)
python/devguide: getting-started/setup-building.rst

If a change is made to Python which relies on some POSIX system-specific
functionality (such as using a new system call), it is necessary to update the

> :file:`configure` script to test for availability of the functionality.

Reviewer: ```suggestion
:cpy-file:`configure` script to test for availability of the functionality.
```

https://github.com/python/devguide/blob/8ffeda5c27586f0c3cfc7653be6fa41a9cc90de0/_extensions/custom_roles.py#L13-L14

This role would be useful, here and below, to link to the file in the main repo.

Author: PTAL

Python's :file:`configure` script is generated from :file:`configure.ac`
using GNU Autoconf.

After editing :file:`configure.ac`, run ``make regen-configure`` to generate
:file:`configure`, :file:`pyconfig.h.in`, and :file:`aclocal.m4`.
When submitting a patch with changes made to :file:`configure.ac`,

> you must also those generated files.

Reviewer 2: ```suggestion
you must also regenerate these files.
```

Author: Thanks! 😄

The recommended and by far the easiest way to regenerate :file:`configure` is::

   $ make regen-configure

This Make target will run a Podman or Docker container (depending on your
environment); you only need to commit the changes.

> Behind the scenes this Make target runs one of the following container images::

Reviewer 2: Is it useful to talk about what the make target does? Easy enough to check for oneself if curious.

Author: It is useful if you're hacking in a cleaned repo (there's no `Makefile` for you to run the `regen-configure` target yet). I could mention that.

Author: Suggestion here: https://github.com/python/devguide/pull/1109/files#r1223569962

Author: PTAL

> This Make target will run a Podman or Docker container (depending on your
> environment); you only need to commit the changes.
> Behind the scenes this Make target runs one of the following container images::

Author: Perhaps:

```suggestion
If you are regenerating :file:`configure` in a clean repo,
run one of the following containers instead::
```

   $ podman run --rm --pull=always -v $(pwd):/src:Z quay.io/tiran/cpython_autoconf:271

>    $ docker run --rm --pull=always -v $(pwd):/src quay.io/tiran/cpython_autoconf:271

Reviewer: If we split each command into its own code block, people can more easily copy just the one they need.

Author: Good point!

If a change is made to Python which relies on some POSIX system-specific
functionality (such as using a new system call), it is necessary to update the
:cpy-file:`configure` script to test for availability of the functionality.
Python's :file:`configure` script is generated from :cpy-file:`configure.ac`

> using GNU Autoconf.

Reviewer 3: ```suggestion
using `GNU Autoconf <https://www.gnu.org/software/autoconf/>`_.
```

Maybe worth explicitly linking it?

After editing :file:`configure.ac`, run ``make regen-configure`` to generate
:file:`configure`, :cpy-file:`pyconfig.h.in`, and :cpy-file:`aclocal.m4`.

> When submitting a patch with changes made to :file:`configure.ac`,

Reviewer 3: ```suggestion
When submitting a pull request with changes made to :file:`configure.ac`,
```

Maybe worth using more modern terminology?

you must also include the changes in these generated files.

> The recommended and by far the easiest way to regenerate :file:`configure` is:

Reviewer 3: ```suggestion
The recommended and by far the easiest way to regenerate :file:`configure` is::
```

If the second suggestion is accepted, also accept this to just use the file-default syntax language, specified to be `console` above.

> .. code-block:: bash

Reviewer 3: This code block (like the others in this file) is `console`, not `bash`, as it includes the `$` prompt. Therefore, you should either change this:

```suggestion
.. code-block:: console
```

Or, like the other code blocks in this file, just use the default language, explicitly set to `console` above, via `::` instead of `.. code-block:: language` (also needs the suggestion to add that):

```suggestion
```

   $ make regen-configure

If you are regenerating :file:`configure` in a clean repo,

> run one of the following containers instead:
> 
> .. code-block:: bash

Reviewer 3: Likewise, either:

```suggestion
run one of the following containers instead:

.. code-block:: console
```

or just

```suggestion
run one of the following containers instead::
```

   $ podman run --rm --pull=always -v $(pwd):/src:Z quay.io/tiran/cpython_autoconf:271

> .. code-block:: bash

Reviewer 3: ```suggestion
.. code-block:: console
```

or

```suggestion
::
```

   $ docker run --rm --pull=always -v $(pwd):/src quay.io/tiran/cpython_autoconf:271

Notice that the images are tagged with ``271``.
Python's :file:`configure.ac` script requires a specific version of
GNU Autoconf.
For Python 3.12 and newer, GNU Autoconf v2.71 is required.
For Python 3.11 and earlier, GNU Autoconf v2.69 is required.
For GNU Autoconf v2.69, change the ``:271`` tag to ``:269``.

> If you cannot, or don't want to use the ``cpython_autoconf`` containers,

Reviewer 3: ```suggestion
If you cannot (or don't want) to use the ``cpython_autoconf`` containers,
```

Avoid awkward/uneven syntax (non-parallel commas)

Author: Thanks! It looks like the closing parenthesis is wrongly placed.

Reviewer 3: Oh, yup :+1:

install the :program:`autoconf-archive` and :program:`pkg-config` utilities,
and make sure the :file:`pkg.m4` macro file located in the appropriate

> :program:`alocal` location:
> 
> .. code-block:: bash

Reviewer 3: Do you mean `aclocal` here?

```suggestion
:program:`aclocal` location:

.. code-block:: console
```

or

```suggestion
:program:`alocal` location::
```

   $ ls $(aclocal --print-ac-dir) | grep pkg.m4

.. note::

   Running :program:`autoreconf` is not the same as running :program:`autoconf`.
   For example, running :program:`autoconf` by itself will not regenerate
   :file:`pyconfig.h.in`.
   :program:`autoreconf` runs :program:`autoconf` and a number of other tools

>    repeatedly as is appropriate.

Reviewer 3: ```suggestion
   repeatedly as appropriate.
```

I believe "as appropriate" is the standard phraseology you're looking for here.
